### PR TITLE
#7809 - Floating-point error when summing Daylighting Zone Fractions

### LINKED ISSUE #7809: Floating-point error when summing Daylighting Zone Fractions
We are running into floating-point rounding errors when setting the fraction of a zone controlled by the daylighting reference.

Although the decimal representation of these fractions given in the IDF sums to exactly 1, it seems that the floating point representation in EnergyPlus is > 1, which causes a Fatal Error.

Would it be possible to tolerate values slightly larger than 1 to accomodate for this case ? Because we are not sure how to proceed when filing the IDF otherwise.

### Details

```
   ** Severe  ** GetDaylightingControls: Fraction of Zone controlled by the Daylighting reference points is > 1.0.
   **   ~~~   ** ..discovered in "Daylighting:Controls" for Zone="Z-F0ON27MII", trying to control 1.00 of the zone.
``` 

Sample Daylighting:Controls that causes EnergyPlus to fail:
```
Daylighting:Controls,
    z-f0on27mii,                   ! Name
    z-f0on27mii,                   ! Zone Name
    splitflux,                     ! Daylighting Method
    on,                            ! Availability Schedule Name
    continuous,                    ! Lighting Control Type
    0.0,                           ! Minimum Input Power Fraction for Continuous or ContinuousOff Dimming Control
    0.0,                           ! Minimum Light Output Fraction for Continuous or ContinuousOff Dimming Control
    1,                             ! Number of Stepped Control Steps
    1.0,                           ! Probability Lighting will be Reset When Needed in Manual Stepped Control
    ,                              ! Glare Calculation Daylighting Reference Point Name
    0.0,                           ! Glare Calculation Azimuth Angle of View Direction Clockwise from Zone y-Axis
    22.0,                          ! Maximum Allowable Discomfort Glare Index
    ,                              ! DElight Gridding Resolution
    z-f0on27mii|jo_ecl_circu+0,    ! Daylighting Reference Point 0 Name
    0.1053,                        ! Fraction of Zone Controlled by Reference Point 0
    200.0,                         ! Illuminance Setpoint at Reference Point 0
    z-f0on27mii|jo_ecl_circu+1,    ! Daylighting Reference Point 1 Name
    0.0936,                        ! Fraction of Zone Controlled by Reference Point 1
    200.0,                         ! Illuminance Setpoint at Reference Point 1
    z-f0on27mii|jo_ecl_circu+2,    ! Daylighting Reference Point 2 Name
    0.1213,                        ! Fraction of Zone Controlled by Reference Point 2
    200.0,                         ! Illuminance Setpoint at Reference Point 2
    z-f0on27mii|jo_ecl_circu+3,    ! Daylighting Reference Point 3 Name
    0.1018,                        ! Fraction of Zone Controlled by Reference Point 3
    200.0,                         ! Illuminance Setpoint at Reference Point 3
    z-f0on27mii|jo_ecl_circu+4,    ! Daylighting Reference Point 4 Name
    0.0893,                        ! Fraction of Zone Controlled by Reference Point 4
    200.0,                         ! Illuminance Setpoint at Reference Point 4
    z-f0on27mii|jo_ecl_circu+5,    ! Daylighting Reference Point 5 Name
    0.0842,                        ! Fraction of Zone Controlled by Reference Point 5
    200.0,                         ! Illuminance Setpoint at Reference Point 5
    z-f0on27mii|jo_ecl_circu+6,    ! Daylighting Reference Point 6 Name
    0.0882,                        ! Fraction of Zone Controlled by Reference Point 6
    200.0,                         ! Illuminance Setpoint at Reference Point 6
    z-f0on27mii|jo_ecl_circu+7,    ! Daylighting Reference Point 7 Name
    0.1026,                        ! Fraction of Zone Controlled by Reference Point 7
    200.0,                         ! Illuminance Setpoint at Reference Point 7
    z-f0on27mii|jo_ecl_circu+8,    ! Daylighting Reference Point 8 Name
    0.1134,                        ! Fraction of Zone Controlled by Reference Point 8
    200.0,                         ! Illuminance Setpoint at Reference Point 8
    z-f0on27mii|jo_ecl_circu+9,    ! Daylighting Reference Point 9 Name
    0.1003,                        ! Fraction of Zone Controlled by Reference Point 9
    200.0;                         ! Illuminance Setpoint at Reference Point 9
```
(note that `1053+0936+1213+1018+0893+0842+0882+1026+1134+1003 = 10000`)

Some additional details for this issue (if relevant):
 - Platform: Ubuntu 18.04
 - Version of EnergyPlus: 9.2

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file here): Post has enough info to produce a unit test
 - [x] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #7809 - Floating-point error when summing Daylighting Zone Fractions
 - Added unit test, then used 
      * a first fix which was an exact solution with 4ULP,
      * then reverted in last commit to something simpler (0.1% tolerance) and I also increased the significant digits in Warning to match otherwise you're getting errors that aren't clear like this one: 

```
   ** Severe  ** GetDaylightingControls: Fraction of Zone controlled by the Daylighting reference points is > 1.0.",
   **   ~~~   ** ..discovered in \"Daylighting:Controls\" for Zone=\"WEST ZONE\", trying to control 1.00 of the zone
```

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/DaylightingManager.cc [hunk 1]
```diff
@@ -5145,6 +5145,12 @@ namespace DaylightingManager {
         int NumAlpha;
         int NumNumber;
 
+        // Smallest deviation from unity for the sum of all fractions
+        // Accept approx 4 to 8 ULP error (technically abs(1.0 + sumFracs) should be close to 2)
+        //   constexpr Real64 FractionTolerance(4 * std::numeric_limits<Real64>::epsilon());
+        // Instead, we use a 0.001 = 0.1% tolerance
+        constexpr Real64 FractionTolerance(0.001);
```
Comment (Myoldmopar) [inline src/EnergyPlus/DaylightingManager.cc:5152]: Works for me.

File: src/EnergyPlus/DaylightingManager.cc [hunk 2]
```diff
@@ -5322,17 +5328,18 @@ namespace DaylightingManager {
                 ErrorsFound = true;
             }
 
-            if (sum(zone_daylight.FracZoneDaylit) < 1.0) {
+            Real64 sumFracs = sum(zone_daylight.FracZoneDaylit);
+            if ( (1.0 - sumFracs) > FractionTolerance) {
```
Comment (Myoldmopar) [inline src/EnergyPlus/DaylightingManager.cc:5332]: That's a lot better.

File: src/EnergyPlus/DaylightingManager.cc [hunk 3]
```diff
@@ -5322,17 +5328,18 @@ namespace DaylightingManager {
...
                 ShowWarningError("GetDaylightingControls: Fraction of Zone controlled by the Daylighting reference points is < 1.0.");
                 ShowContinueError("..discovered in \"" + cCurrentModuleObject + "\" for Zone=\"" + cAlphaArgs(2) + "\", only " +
-                                  RoundSigDigits(sum(zone_daylight.FracZoneDaylit), 2) + " of the zone is controlled.");
-            }
-            if (sum(zone_daylight.FracZoneDaylit) > 1.0) {
+                        RoundSigDigits(sum(zone_daylight.FracZoneDaylit), 3) + " of the zone is controlled.");
```
Comment (Myoldmopar) [inline src/EnergyPlus/DaylightingManager.cc:5335]: And increasing this makes good sense too.

File: tst/EnergyPlus/unit/DaylightingManager.unit.cc [hunk 4]
```diff
@@ -1395,3 +1398,274 @@ TEST_F(EnergyPlusFixture, DaylightingManager_DayltgInteriorIllum_Test)
...
+    }
+
+    // It does sum up to 1.0
+    EXPECT_DOUBLE_EQ(1.0, sum);
+    EXPECT_FALSE(std::abs(sum - 1.0) > std::numeric_limits<double>::epsilon());
+    // Yet, if you are being very litteral, then it's slightly more
+    EXPECT_TRUE(sum > 1.0);
+    EXPECT_FALSE(sum < 1.0);
+
+}
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/DaylightingManager.unit.cc:1595]: This unit tests clearly demonstrates the equal to one case.

File: tst/EnergyPlus/unit/DaylightingManager.unit.cc [hunk 5]
```diff
@@ -1395,3 +1398,274 @@ TEST_F(EnergyPlusFixture, DaylightingManager_DayltgInteriorIllum_Test)
...
+
+TEST_F(EnergyPlusFixture, DaylightingManager_GetInputDaylightingControls_NotAroundOne)
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/DaylightingManager.unit.cc:1597]: And this one clearly demonstrates the proper warning message with the extra digit printed.  Good stuff.

**Review by Myoldmopar — APPROVED** [review]
This is a great fix and set of unit tests.  I'll pull it locally and verify.

**Comment by Myoldmopar** [comment]
I didn't find a defect file to check out, but the unit tests here are really great test cases.  I pulled in develop to this branch and rebuilt and reran.  All unit tests pass.  I'm calling this done.  Thanks @jmarrec.

**Comment by jmarrec** [comment]
Thanks for the review!

**Pull request merged**